src/main/kotlin/org/sert2521/deepspace/climber/ClimberCommands.kt

```diff
@@ -1,24 +1,110 @@
 package org.sert2521.deepspace.climber
 
 import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import org.sert2521.deepspace.drivetrain.Drivetrain
+import org.sert2521.deepspace.util.PIDFController
+import org.sert2521.deepspace.util.timer
+import org.sert2521.deepspace.util.tol
+import org.team2471.frc.lib.coroutines.MeanlibDispatcher
 import org.team2471.frc.lib.coroutines.parallel
 import org.team2471.frc.lib.coroutines.periodic
+import org.team2471.frc.lib.coroutines.suspendUntil
 import org.team2471.frc.lib.framework.use
 
-suspend fun Climber.driveUntilAbovePlatform() = use(this) {
+private suspend fun Climber.elevateWithPidTo(state: ClimberState, front: Boolean = true, rear: Boolean = true) {
+    val frontPid = PIDFController(kp = 0.6, offset = 0.29)
+    val rearPid = PIDFController(kp = 0.67, offset = 0.221)
     periodic {
-        if (elevated) drive(Climber.DriveDirection.FORWARD) else this@periodic.stop()
+        if (front) setFrontSpeed(frontPid.update(state.position, frontLegPosition))
+        if (rear) setRearSpeed(rearPid.update(state.position, rearLegPosition))
+        (frontLegPosition in (state.position tol ALLOWED_CLIMBER_ERROR))
+        (rearLegPosition in (state.position tol ALLOWED_CLIMBER_ERROR))
     }
-    stop()
 }
 
-suspend fun Climber.elevateTo(position: Int) = use(this) {
-    ensureSynced(Climber.SyncDirection.FRONT_TO_REAR)
-    GlobalScope.parallel(
-        { elevateFrontTo(position) },
-        { elevateRearTo(position) }
-    )
-    ensureSynced(Climber.SyncDirection.FRONT_TO_REAR)
+suspend fun ClimberDrive.driveForward() = use(this) {
+    try {
+        periodic {
+            ClimberDrive.drive(ClimberDrive.DriveDirection.FORWARD)
+        }
+    } finally {
+        stop()
+    }
 }
 
-suspend fun Climber.elevateToTop() = elevateTo(CLIMBER_TOP_SENSOR_POSITION)
+suspend fun runClimbSequence(state: ClimberState) {
```

Comment:
need to test, but I'm guessing this will probably be better if we get rid of all the inner `use` and instead do:

```suggestion
suspend fun runClimbSequence(state: ClimberState) = use(Climber, ClimberDrive, Drivetrain)
```

That way it'll be much easier to cancel if need be.

Reply:
Also, meanlib now returns a CoroutineScope from `use` so we can stop using GlobalScope in these bodies.